File: linkedlist/singlylinkedlist/singlylinkedlist.js

/*
		Time complexity :

			access    : O(n)
			delete    : O(1)
			insertion : O(1)
			search    : O(n) 
*/

class node
{
	constructor(data = null)
	{
		this.data = data
		this._next = null ;
	}
}

class linkedlist
{
	constructor()
	{
		this.head = new node() ;
	}

	_display()
	{
		let array = [] ;
		let temp = this.head ;
		while(temp._next !== null)
		{
			temp = temp._next ;
			array.push(temp.data) ;
		}
		return array ;
	}

	_append(value)
	{
		let newnode = new node(value) ;
		if(this.head === null)
		{
			this.head = newnode ;
			return ;
		}
		let temp = this.head ;
		while(temp._next !== null)
		{
			temp = temp._next ;
		}
		temp._next = newnode ;
	}

	_size()
	{
		let temp = this.head ;
		let s = 0 ;

		while(temp._next !== null)
		{
			temp = temp._next ;
			s++ ;
		}
		return s ;
	}

	_insert(index , value)
	{
		let newnode = new node(value) ;
		if(this.head === null)
		{
			this.head = newnode ;
			return ;
		}
		let temp = this.head ;
		for(let i = 0 ; i < index ; i++)
		{
			if( temp.next != null)
				break ;
			else
				temp = temp._next ;
		}
		newnode._next = temp._next ;
		temp._next = newnode ;
	}

	_remove(value)
	{
		if(this.head === null)
			return ;

		let prev = this.head ;
		let temp = this.head ;

		if(this.head.data == value)
		{
			this.head = this.head._next ;

		}
		else
		{
			while(temp.data != value && temp._next != null)
			{
				prev = temp ;
				temp = temp._next ;
			}
			if(temp == null)
				return ;
			prev._next = temp._next ;
		}
	}

	_removeat(index)
	{
		if(this.head === null)
			return ;
		if(index === 0)
		{
			this.head = this.head._next ;
			return ;
		}
		let temp = this.head ;

		for(let i = 0 ; i < index && temp._next != null ; i++)
		{
			temp = temp._next
		}
		if(temp._next === null)
			return ;
		temp._next = temp._next._next ;
	}

	_reverse()
	{
		if(this.head === null)
			return ;
		let prev = null ;
		let curr = this.head._next ;
		let _next = null ;

		while(curr != null)
		{
			_next = curr._next ;
			curr._next = prev ;
			prev = curr ;
			curr = _next ;
		}
		this.head._next = prev ;
	}

}

let l = new linkedlist() ;

l._append("A") ;
l._append("B") ;
l._append("C") ;
l._append("D") ;
l._append("E") ;
l._insert(1 , "F") ;
l._insert(6, "G") ;
l._remove("C") ;
l._removeat(5) ;
l._reverse() ;

console.log(l._display()) ;
console.log(l._size());

File: linkedlist/singlylinkedlist/singlylinkedlist.py

'''
		Time complexity :

			access    : O(n)
			delete    : O(1)
			insertion : O(1)
			search    : O(n) 
'''

class node :
	def __init__(self , data = None) :
		self.data = data
		self._next = None


class linkedlist :

	def __init__(self) :
		self.head = node()

	def _append(self , value) :

		newnode = node(value)
		if(self.head == None) :
			self.head = value
			return
		temp = self.head
		while(temp._next != None) :
			temp = temp._next
		temp._next = newnode

	def _display(self) :

		array = []
		temp = self.head
		while(temp._next != None) :
			temp = temp._next
			array.append(temp.data)
		return array

	def _size(self) :

		s = 0
		temp = self.head
		while(temp._next != None) :
			s += 1
			temp = temp._next
		return s

	def _insert(self , index , value) :

		newnode = node(value)
		if(self.head == None) :
			self.head = newnode
		else :
			temp = self.head
			for i in range(0 , index) :
				if temp._next == None :
					break
				else :
					temp = temp._next
			newnode._next = temp._next
			temp._next = newnode


	def _remove(self , value) :

		if self.head == None :
			return
		
		temp = self.head
		prev = self.head

		if self.head.data == value :
			self.head = self.head.next
		else :
			while(temp.data != value and temp._next != None) :
				prev = temp
				temp = temp._next

			if temp == None :
				return
			prev._next = temp._next

	def _removeat(self , index) :

		if self.head == None :
			return
		if index == 0 :
			self.head = self.head._next
		temp = self.head
		for i in range(0 , index) :
			if temp._next == None :
				break
			else :
				temp = temp._next
		if temp._next == None :
			return
		temp._next = temp._next._next

	def _reverse(self) :

		if self.head == None :
			return
		prev = None
		curr = self.head._next
		_next = None

		while curr != None :
			_next = curr._next
			curr._next = prev
			prev = curr
			curr = _next
		self.head._next = prev


l = linkedlist()

l._append("A") ;
l._append("B") ;
l._append("C") ;
l._append("D") ;
l._append("E") ;
l._insert(1 , "F") ;
l._insert(6, "G") ;
l._remove("C") ;
l._removeat(5) ;
l._reverse() ;

print(l._display())
print(l._size())